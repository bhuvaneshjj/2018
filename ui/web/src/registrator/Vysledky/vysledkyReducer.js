import moment from 'moment';
import convertDuration from 'ui-common/convertDuration';
import { AKTUALNI_ROK, TYPY_KATEGORII } from '../../constants';
import { findDokonceno } from '../../Util';
import { getKategorieProTyp } from '../../entities/rocniky/rocnikyReducer';
import { getPoradiSorted } from '../Poradi/poradiReducer';

const months = [
  'ledna',
  'února',
  'března',
  'dubna',
  'května',
  'června',
  'července',
  'srpna',
  'září',
  'října',
  'listopadu',
  'prosince'
];

const pohlaviMnozne = { muž: 'muži', žena: 'ženy' };

const popisekKategorie = ({ pohlavi, vek, typ }) => {
  if (!pohlavi) {
    return typ.toUpperCase().substring(0, 1);
  }
  if (!vek) {
    return pohlaviMnozne[pohlavi];
  }
  if (vek.max === 150) {
    return `${pohlaviMnozne[pohlavi]} ${vek.min} a více`;
  }
  return `${pohlaviMnozne[pohlavi]} ${vek.min}-${vek.max}`;
};

const popisekTypu = {
  maraton: 'maratonci',
  půlmaraton: 'půlmaratonci',
  cyklo: 'cyklisté',
  koloběžka: 'koloběžky',
  pěší: 'turisté'
};

export const getVysledky = ({ rocniky, ucastnici, rok = AKTUALNI_ROK }) => {
  const rocnik = rocniky.byRoky[rok];
  const date = new Date(rocnik.datum);

  const vysledky = {
    title: 'Výsledková listina',
    subtitle: `Jirkovský crossmarathon a Jirkovský půlmarathon ${rok}`,
    datum: `${date.getDate()}. ${months[date.getMonth()]} ${date.getFullYear()}`,
    summary: { startovalo: 0, dokoncilo: 0 },
    typy: {}
  };

  TYPY_KATEGORII.forEach(typ => {
    const stats = { startovalo: 0, dokoncilo: 0 };
    const poradi = getPoradiSorted({
      kategorieFilter: typ,
      sortColumn: 'absPoradi',
      rocniky,
      ucastnici,
      rok
    });

    const ucastniciProTyp = poradi.map(
      ({
        id,
        prijmeni,
        jmeno,
        narozeni,
        obec,
        stat,
        klub,
        kategorie,
        startCislo,
        dokonceno,
        cas,
        absPoradi,
        relPoradi
      }) => {
        const { zkratka } = kategorie;
        let casAsText;
        if (dokonceno) {
          const duration = convertDuration(moment.duration(cas));
          casAsText = `${duration.hours}:${duration.mins}:${duration.secs},${duration.subsecs}`;
        } else {
          casAsText = findDokonceno(dokonceno).popisek;
        }
        const ucastnik = {
          id,
          prijmeni,
          jmeno,
          narozeni: narozeni.rok,
          misto: `${klub || obec}${stat !== 'Česká republika' ? ` (${stat})` : ''}`,
          kategorie,
          startCislo,
          dokonceno,
          cas: casAsText,
          absPoradi,
          relPoradi
        };

        if (!stats[zkratka]) {
          stats[zkratka] = {
            startovalo: 0,
            dokoncilo: 0,
            popisek: popisekKategorie(kategorie),
            zkratka
          };
        }
        stats[zkratka].startovalo += 1;
        stats.startovalo += 1;
        vysledky.summary.startovalo += 1;
        if (dokonceno) {
          stats[zkratka].dokoncilo += 1;
          stats.dokoncilo += 1;
          vysledky.summary.dokoncilo += 1;
        }

        return ucastnik;
      }
    );

    const kategorieProTyp = getKategorieProTyp({ typ, rocniky, rok });
    const vsechnyZkratky = kategorieProTyp.list.map(({ zkratka }) =>
      stats[zkratka] ? zkratka : undefined
    );
    const zkratky = vsechnyZkratky.filter(zkratka => zkratka);

    vysledky.typy[typ] = {
      popisek: popisekTypu[typ],
      startCisla: !!rocnik.kategorie[typ].startCisla,
      stats,
      typ,
      ucastnici: ucastniciProTyp,
      zkratky
    };
  });

  return vysledky;
};
