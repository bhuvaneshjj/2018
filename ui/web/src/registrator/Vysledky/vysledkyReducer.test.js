import ucastniciTestData from '../../entities/ucastnici/ucastniciTestData';
import { getVysledky } from './vysledkyReducer';

it('getVysledky() - test db', () => {
  const vysledky = getVysledky({ ...ucastniciTestData.entities });

  expect(vysledky).toMatchSnapshot();
});
