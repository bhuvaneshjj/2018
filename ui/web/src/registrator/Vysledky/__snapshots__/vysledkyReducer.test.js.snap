// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getVysledky() - test db 1`] = `
Object {
  "datum": "8. června 2019",
  "subtitle": "Jirkovský crossmarathon a Jirkovský půlmarathon 2019",
  "summary": Object {
    "dokoncilo": 2,
    "startovalo": 3,
  },
  "title": "Výsledková listina",
  "typy": Object {
    "cyklo": Object {
      "popisek": "cyklisté",
      "startCisla": true,
      "stats": Object {
        "dokoncilo": 0,
        "startovalo": 0,
      },
      "typ": "cyklo",
      "ucastnici": Array [],
      "zkratky": Array [],
    },
    "koloběžka": Object {
      "popisek": "koloběžky",
      "startCisla": true,
      "stats": Object {
        "dokoncilo": 0,
        "startovalo": 0,
      },
      "typ": "koloběžka",
      "ucastnici": Array [],
      "zkratky": Array [],
    },
    "maraton": Object {
      "popisek": "maratonci",
      "startCisla": true,
      "stats": Object {
        "dokoncilo": 0,
        "startovalo": 0,
      },
      "typ": "maraton",
      "ucastnici": Array [],
      "zkratky": Array [],
    },
    "pěší": Object {
      "popisek": "turisté",
      "startCisla": false,
      "stats": Object {
        "dokoncilo": 0,
        "startovalo": 0,
      },
      "typ": "pěší",
      "ucastnici": Array [],
      "zkratky": Array [],
    },
    "půlmaraton": Object {
      "popisek": "půlmaratonci",
      "startCisla": true,
      "stats": Object {
        "1Ž": Object {
          "dokoncilo": 2,
          "popisek": "ženy 18-39",
          "startovalo": 2,
          "zkratka": "1Ž",
        },
        "4M": Object {
          "dokoncilo": 0,
          "popisek": "muži 60 a více",
          "startovalo": 1,
          "zkratka": "4M",
        },
        "dokoncilo": 2,
        "startovalo": 3,
      },
      "typ": "půlmaraton",
      "ucastnici": Array [
        Object {
          "absPoradi": 1,
          "cas": "2:06:32,60",
          "dokonceno": true,
          "id": "7a09b1fd371dec1e99b7e142",
          "jmeno": "Hana",
          "kategorie": Object {
            "id": "5a587e1b051c181132cf83d9",
            "pohlavi": "žena",
            "typ": "půlmaraton",
            "vek": Object {
              "max": 39,
              "min": 18,
            },
            "zkratka": "1Ž",
          },
          "misto": "SK Nudle",
          "narozeni": 1999,
          "prijmeni": "Zralá",
          "relPoradi": 1,
          "startCislo": 11,
        },
        Object {
          "absPoradi": 2,
          "cas": "3:15:32,60",
          "dokonceno": true,
          "id": "9ccbc71dedc1e99b7e1d671",
          "jmeno": "Božena",
          "kategorie": Object {
            "id": "5a587e1b051c181132cf83d9",
            "pohlavi": "žena",
            "typ": "půlmaraton",
            "vek": Object {
              "max": 39,
              "min": 18,
            },
            "zkratka": "1Ž",
          },
          "misto": "Bor",
          "narozeni": 2001,
          "prijmeni": "Půlmaratonka",
          "relPoradi": 2,
          "startCislo": 8,
        },
        Object {
          "absPoradi": undefined,
          "cas": "nedokončeno",
          "dokonceno": false,
          "id": "5a09b1fd371dec1e99b7e1c9",
          "jmeno": "Roman",
          "kategorie": Object {
            "id": "5a587e1b051c181132cf83d7",
            "pohlavi": "muž",
            "typ": "půlmaraton",
            "vek": Object {
              "max": 150,
              "min": 60,
            },
            "zkratka": "4M",
          },
          "misto": "Ostrava 2",
          "narozeni": 1956,
          "prijmeni": "Balabák",
          "relPoradi": undefined,
          "startCislo": 15,
        },
      ],
      "zkratky": Array [
        "4M",
        "1Ž",
      ],
    },
  },
}
`;
